### DISCUSSION TITLE
Creating a Representation of Electrical Cables with IFCopenshell - Assigning Geometry Data Woes

### DISCUSSION THREAD

**Discussion by OrionSehn** [IfcOpenShell/IfcOpenShell#3260] [Q&A] [answered]
# Creating a Representation of Electrical Cables with IFCopenshell

My goal is to draw a representation of an electrical cable in an IFC file, and I'm having trouble reading through the documentation (as the python user guide detailing creating entities has not been [completed.](https://blenderbim.org/docs-python/ifcopenshell-python/developer_guide.html
)

There are detailed examples for how to create [walls and doors](https://blenderbim.org/docs-python/ifcopenshell-python/code_examples.html#create-a-simple-model-from-scratch), but there are not examples on creating/assigning geometry to other objects. The following Is the code I have tried:

```
import ifcopenshell
import ifcopenshell.geom
from ifcopenshell.api import run

# Create a blank model
model = ifcopenshell.file()
project = run("root.create_entity", model, ifc_class="IfcProject", name="My Project")
run("unit.assign_unit", model)
context = run("context.add_context", model, context_type="Model")
body = run("context.add_context", model, context_type="Model",
    context_identifier="Body", target_view="MODEL_VIEW", parent=context)

# Let's create a cable
cable = run("root.create_entity", model, ifc_class="IfcCableSegment")

representation = run("geometry.add_representation", model, context=body, length=5, height=3, thickness=0.2)

# Add a new cable-like body geometry, from sample cable
run("geometry.assign_representation", model, product=cable, representation=representation)

# Write out to a file
model.write(r"C:\Users\OTSehn\ProjectData\Cable-Opt\created_example-jun.ifc")

```

I figured I could create an IfcCableSegment, then assign the geometry data to this, but when I try, I get this error 

```
Traceback (most recent call last):
  File "c:\Users\OTSehn\Projects\Cable-Tray-Opt\src\IFCConvert-Sample.py", line 16, in <module>
    representation = run("geometry.add_representation", model, context=body, length=5, height=3, thickness=0.2)
  File "C:\Users\OTSehn\Projects\Cable-Tray-Opt\env\lib\site-packages\ifcopenshell\api\__init__.py", line 61, in run
    importlib.import_module(f"ifcopenshell.api.{usecase_path}")
  File "C:\Users\OTSehn\AppData\Local\Programs\Python\Python310\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Users\OTSehn\Projects\Cable-Tray-Opt\env\lib\site-packages\ifcopenshell\api\geometry\add_representation.py", line 19, in <module>
    import bpy
ModuleNotFoundError: No module named 'bpy'

```

This is not a dependency for the IfcOpenshell project. Even when I install this with pip, I get a different package missing error:

```

Traceback (most recent call last):
  File "c:\Users\OTSehn\Projects\Cable-Tray-Opt\src\IFCConvert-Sample.py", line 16, in <module>
    representation = run("geometry.add_representation", model, context=body, length=5, height=3, thickness=0.2)
  File "C:\Users\OTSehn\Projects\Cable-Tray-Opt\env\lib\site-packages\ifcopenshell\api\__init__.py", line 61, in run
    importlib.import_module(f"ifcopenshell.api.{usecase_path}")
  File "C:\Users\OTSehn\AppData\Local\Programs\Python\Python310\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Users\OTSehn\Projects\Cable-Tray-Opt\env\lib\site-packages\ifcopenshell\api\geometry\add_representation.py", line 23, in <module>
    from blenderbim.bim.module.geometry.helper import Helper
  File "C:\Users\OTSehn\AppData\Roaming\Blender Foundation\Blender\3.5\scripts\addons\blenderbim\__init__.py", line 40, in <module>
    import blenderbim.bim
  File "C:\Users\OTSehn\AppData\Roaming\Blender Foundation\Blender\3.5\scripts\addons\blenderbim\bim\__init__.py", line 22, in <module>
    from . import handler, ui, prop, operator, helper
  File "C:\Users\OTSehn\AppData\Roaming\Blender Foundation\Blender\3.5\scripts\addons\blenderbim\bim\handler.py", line 23, in <module>
    import blenderbim.tool as tool
  File "C:\Users\OTSehn\AppData\Roaming\Blender Foundation\Blender\3.5\scripts\addons\blenderbim\tool\__init__.py", line 46, in <module>
    from blenderbim.tool.resource import Resource
  File "C:\Users\OTSehn\AppData\Roaming\Blender Foundation\Blender\3.5\scripts\addons\blenderbim\tool\resource.py", line 33, in <module>
    import ifcopenshell.util.resource
ModuleNotFoundError: No module named 'ifcopenshell.util.resource'

```

This module is not in the [documentation](https://blenderbim.org/docs-python/autoapi/ifcopenshell/util/index.html) and does not seem to exist yet. 

If you exclude the last two lines of code, it does create a cable object, at (0,0,0) without any attributes.

```
import ifcopenshell
import ifcopenshell.geom
from ifcopenshell.api import run

# Create a blank model
model = ifcopenshell.file()
project = run("root.create_entity", model, ifc_class="IfcProject", name="My Project")
run("unit.assign_unit", model)
context = run("context.add_context", model, context_type="Model")
body = run("context.add_context", model, context_type="Model",
    context_identifier="Body", target_view="MODEL_VIEW", parent=context)

# Let's create a cable
cable = run("root.create_entity", model, ifc_class="IfcCableSegment")

# representation = run("geometry.add_representation", model, context=body, length=5, height=3, thickness=0.2)

# Add a new cable-like body geometry, from sample cable
# run("geometry.assign_representation", model, product=cable, representation=representation)

# Write out to a file
# model.write(r"C:\Users\OTSehn\ProjectData\Cable-Opt\created_example-jun.ifc")
```

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/97123736/4345f243-659b-440c-a7e8-bfddb0757653)

I'm a little at a loss of how to create objects to represent cables.
Can somebody please point me in the right direction?

# TLDR 

- How do you create an entity to represent and electrical cable? Can anybody provide me with a code snippet or further documentation?
- Can anybody give me some resources on what IFC objects exist, and when to use what resource?
- The python geometry.add_representation seems to require packages or portions of the code which do not exist.
- Can you please put an effort into creating some additional examples of how you can create objects, and assign xyz values to them (this is not included in the [current code examples](https://blenderbim.org/docs-python/ifcopenshell-python/code_examples.html#create-a-simple-model-from-scratch))
- I would appreciate any advice
- I did search for similar discussions

Orion

**Comment by arunkvijay** [comment]
Hi Orion, I can see that [geometry.add_representation](https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/geometry/add_representation.py#L32-L58) depends on blender's data model and blender object. Not sure how the workflow is though. Would love to explore this further.
Arun

**Comment by aothms** [comment]
Indeed, while most of the API is indeed agnostic to blender, for the geometry-heavy bits, creating a proper layer of isolation between would be challenging. Any help on this front is appreciated, but I think for now it is simplest to just use the lower level direct instance manipulation and create a representation manually.

- copy https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/geometry/add_representation.py#L287-L293
- with Items=f.createIfcSweptDiskSolid(...)
- call assign_representation from api.

> ModuleNotFoundError: No module named 'ifcopenshell.util.resource'

https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/util/resource.py

It does seem to exist, but probably is recent. IfcOpenShell on pip is sometimes a bit outdated. Maybe copy manually from here https://github.com/IfcOpenBot/IfcOpenShell/commit/b33e58f40b628bb892bc5d987575917453c7bde7#comments

**Reply by Moult** [reply]
> creating a proper layer of isolation between would be challenging

The ShapeBuilder utility is a slowly growing utility that helps solve this. To date, we use it to generate project asset libraries like furniture, railings, stairs, windows, doors, and other fun things so I'd say it's pretty capable for things you can represent with simple extrusions. It's heavily inspired by the capabilities of OpenSCAD which offers programmatic CAD. @OrionSehn if you use it and have comments, it will help the utility mature faster!

This is a file that might be of interest: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/scripts/generate_furniture_library.py

Some example functions of it in use:

 - 2D bed: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/scripts/generate_furniture_library.py#L310-L326
 - 3D bed: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/scripts/generate_furniture_library.py#L328-L330
 - 2D and 3D chair: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/scripts/generate_furniture_library.py#L383-L450
 - 3D toilet: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/scripts/generate_furniture_library.py#L1155

**Comment by Moult** [comment, marked as answer]
As @aothms said, the majority of the API is agnostic to Blender. However the `add_representation` API depends on Blender. This is a bit historic and we should really rename it to `add_blender_representation` to prevent confusion.

There are hundreds of ways to create geometry. The API only handles the most common ways like parametric layered and profile extrusions and arbitrary meshes (but maybe as you say "cable-like" or "rebar-like" geometry should be added to the API for convenience).

Once you start deviating and wanting to add your own geometry, you have two options: either go down to the core level and create swept disk solids etc from scratch as @aothms is suggesting, or alternatively, and this is what I'd recommend: use the ShapeBuilder utility instead, which is an API somewhat in-between the high-level API and the low-level total manual approach. This is a solution using the ShapeBuilder utility:

```python
import ifcopenshell
import ifcopenshell.geom
import ifcopenshell.util.shape_builder
from ifcopenshell.api import run

# Create a blank model
model = run("project.create_file", version="IFC4")
project = run("root.create_entity", model, ifc_class="IfcProject", name="My Project")
run("unit.assign_unit", model)
context = run("context.add_context", model, context_type="Model")
body = run("context.add_context", model, context_type="Model",
    context_identifier="Body", target_view="MODEL_VIEW", parent=context)

site = run("root.create_entity", model, ifc_class="IfcSite", name="My Site")
run("aggregate.assign_object", model, relating_object=project, product=site)

# Let's create a cable
cable = run("root.create_entity", model, ifc_class="IfcCableSegment")

sb = ifcopenshell.util.shape_builder.ShapeBuilder(model)
curve = sb.polyline([(0., 0., 0.), (1000., 0., 0.), (1707., 293., 0.), (2000., 1000., 0.), (2000., 2000., 0.)], arc_points=[2])
swept_curve = sb.create_swept_disk_solid(curve, 5) # 5 is radius
representation = sb.get_representation(body, swept_curve)

# Add a new cable-like body geometry, from sample cable
run("geometry.assign_representation", model, product=cable, representation=representation)

run("spatial.assign_container", model, relating_structure=site, product=cable)

# Write out to a file
model.write("/home/dion/test.ifc")
```

It produces this:

![2023-06-07-171959_773x329_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/7c8d4265-025c-42c2-8e96-215a27c824ab)

> Can anybody give me some resources on what IFC objects exist, and when to use what resource?

That's what the IFC documentation is for, and you can see a list of IFC entities here: https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/annex-c.html - I'd also say our API docs do a pretty good job of explaining things too: https://blenderbim.org/docs-python/autoapi/ifcopenshell/api/index.html which is less technical and comes with lots of code examples. Do you have any specific questions?

> Can you please put an effort into creating some additional examples of how you can create objects, and assign xyz values to them (this is not included in the [current code examples](https://blenderbim.org/docs-python/ifcopenshell-python/code_examples.html#create-a-simple-model-from-scratch))

Unfortunately writing docs about creating geometry takes a lot of effort. There's a lot of functionality, just a bit hidden and constantly being updated. Our docs have improved drastically in the past year, and I suspect over time we'll get even better docs dedicated to geometry. Would you like to help?

**Reply by OrionSehn** [reply]
Hello!

Thank you so much for the detailed and prompt response! This is exactly what I was hoping for, and seems to be detailed enough for my team's purposes. I really like the approach of using the shapebuilder.

I also want to thank you so much for all of your documentation references, I appreciate them so much, I think having a related works/further reading section at the end of your documenation would make this path out much clearer for others. As for helping add to your documentation, **I can commit some time to helping build documentation once I get a little more familiar with the package**, and I would like to leave the metaphorical path I'm walking on better than I found it. 


### Your Solution:
This is exactly what I was looking for, and worked exactly as I had hoped. 

One note, I do not think mathutils is in your current build dependencies for a pip install, and I had to go install the Microsoft Visual C++ Build Tools.
https://pypi.org/project/mathutils/

```
Traceback (most recent call last):
  File "c:\Users\OTSehn\Projects\Cable-Tray-Opt\src\IFCConvert.py", line 15, in <module>
    import ifcopenshell.util.shape_builder
  File "C:\Users\OTSehn\Projects\Cable-Tray-Opt\env\lib\site-packages\ifcopenshell\util\shape_builder.py", line 23, in <module>
    from mathutils import Vector, Matrix
ModuleNotFoundError: No module named 'mathutils'
```
I updated to the latest version on pip (released may.23 https://pypi.org/project/ifcopenshell/0.7.0.230523/) 
I'm not sure if I'm using a wrong version, or if these tools have yet to have been pushed into the build? 
It did work after I installed this.

Thank you so so much for your help.

**Reply by Moult** [reply]
You're right, mathutils is not in the current build dependencies. I think we should offer it as an "extra" (in pip you can offer "extra" flags that pull in more deps).  Feel free to file a bug about this so we don't forget. Or better yet, a PR to the pip metadata.

**Reply by aothms** [reply]
> I can commit some time to helping build documentation once I get a little more familiar with the package

That would be greatly appreciated :)
